src/julia_gc.c

```diff
@@ -448,11 +457,51 @@ void InitMarkFuncBags(UInt type, TNumMarkFuncBags mark_func)
     TabMarkFuncBags[type] = mark_func;
 }
 
+static inline int JMarkGapObjSafe(void * obj)
+{
+    // only traverse objects internally used by GAP
+    void *ty = jl_typeof(obj);
+    if (ty != datatype_mptr && ty != datatype_bag
+        && ty != datatype_largebag && ty != jl_weakref_type)
```

Comment:
There are three calls to `JMarkGapObjSafe`. Two are in `MarkBag` -- but for that caller, we only need to accept `datatype_mptr` and `jl_weakref_type`. And if we added a custom `MarkBagWeakly` function (to mirror the name used in GASMAN) which only accepts `jl_weakref_type`, we could reduce the check to `jl_typeis(obj, datatype_mptr)` which is more efficient.

And `MarkBagWeakly` could potentially be considerably simpler than `MarkBag`, as (a) it is very rarely used, and (b) we control where it is called from tightly, and so we'd not have to validate as much, and could skipping the `MarkCache`. So this might suffice:
```C
void MarkBagWeakly(Bag bag)
{
    if (bag == 0)
        return;
    jl_value_t * p = (jl_value_t *)bag;
    GAP_ASSERT(jl_typeis(p, jl_weakref_type));
    if (jl_gc_mark_queue_obj(JuliaTLS, p))
        YoungRef++;
}
```


Anyway, that leaves the third call to `JMarkGapObjSafe`, which is in `TryMark`. Note that in the `SCAN_STACK_FOR_MPTRS_ONLY` case, it also only checks `jl_typeis(obj, datatype_mptr)`.

So, let's change `JMarkGapObjSafe` to only check for `datatype_mptr`, and then replace the one remaining call to it with a copy of the code that's now here. And then we could rename this function to something more specific... I want to say `JMarkMPtr`, but we already use that name (which, BTW, I find rather confusing -- `JMarkMPtr` is a Julia marking function, while `JMark` is a helper function for use in lots of places

Reply:
In `TryMark()`, `JMarkGapObjSafe()` is not called if `SCAN_STACK_FOR_MPTRS_ONLY` is enabled. Otherwise, it's used in lieu of the MPtr type check and needs to cover all cases.

I did not specialize the type check in `JMarkGapObjSafe()` further for other use cases, because it will succeed on the first hit for almost all cases with `SCAN_STACK_FOR_MPTRS_ONLY` enabled and is probably an unnecessary optimization.

Reply:
I am on board with renaming the functions, though. That thought had occurred to me, too.